### PR TITLE
Clean up / refactor logging

### LINKED ISSUE #485: seterrors should log as errors
Now (v2.4) that we are no longer throwing an error when logging, please log errors when we chunk.setError and stream emit error

https://github.com/linkedin/dustjs/blob/master/lib/dust.js#L448
https://github.com/linkedin/dustjs/blob/master/lib/dust.js#L472

There may be other places as well.

not sure whether this should be logged as a warning or an error
https://github.com/linkedin/dustjs/blob/master/lib/dust.js#L359

### PR DESCRIPTION
**Pull request by sethkinast**
Similar to linkedin/dustjs-helpers#119

Normalizes all log messages, removes some spew, fixes some logs that would never be logged, and removes the indexOf shim in favor of a map lookup.

Closes #485

### CODE REVIEW

**Comment by sethkinast** [comment]
Reduces logspew by ~600 lines

File: lib/dust.js [hunk 1]
```diff
@@ -115,9 +115,14 @@
     return dust.compileFn(source)(context, callback);
   };
 
+  /**
+   * Compile a template to an invokable function.
+   * If `name` is provided, also registers the template under `name`.
+   * @param source {String} template source
+   * @param name {String} template name
```
Comment (jimmyhchan) [inline lib/dust.js]: template name is optional. i think there's a jsdoc syntax for that  `@param [name] {string}`

File: lib/dust.js [hunk 2]
```diff
@@ -390,24 +365,30 @@
   };
 
   Context.prototype.getBlock = function(key, chk, ctx) {
+    var tempChk = new Chunk(),
```
Comment (jimmyhchan) [inline lib/dust.js]: not sure we need to hoist the Chunk invocation as well.  nit pick

File: lib/dust.js [hunk 3]
```diff
@@ -488,62 +469,52 @@
...
     } else {
-      this.events[type].push(callback);
+      handlers.push(callback);
     }
     return this;
   };
 
   Stream.prototype.pipe = function(stream) {
-    this.on('data', function(data) {
+    return this
```
Comment (jimmyhchan) [inline lib/dust.js]: NICE!

File: lib/dust.js (hunk 2 above)
Comment (sethkinast) [inline lib/dust.js]: Don't even need the var at all in fact, fixed thanks

File: lib/dust.js (hunk 1 above)
Comment (sethkinast) [inline lib/dust.js]: Cool, updated

**Comment by jimmyhchan** [comment]
LOOKS good!

**Comment by sethkinast** [comment]
Spew comparison:

Old: https://travis-ci.org/linkedin/dustjs/jobs/55233215
New: https://travis-ci.org/linkedin/dustjs/jobs/55252819

File: lib/dust.js [hunk 4]
```diff
@@ -710,7 +677,6 @@
     } else if (skip) {
       return skip(this, context);
     }
-    dust.log('Not rendering exists (?) block in template [' + context.getTemplateName() + '], because above key was not found', DEBUG);
```
Comment (prashn64) [inline lib/dust.js]: We may have to keep this one.

File: lib/dust.js [hunk 5]
```diff
@@ -783,11 +744,11 @@
       try {
         return dust.helpers[name](chunk, context, bodies, params);
       } catch(e) {
```
Comment (prashn64) [inline lib/dust.js]: e -> err

**Comment by sethkinast** [comment]
Updated with Prash's comments:

* e -> err
* returned debug messages for exists and nexists

File: lib/dust.js (hunk 4 above)
Comment (sethkinast) [inline lib/dust.js]: added back

**Pull request merged**